src/Microsoft.AspNet.Mvc.ModelBinding/Validation/CompareAttributeAdapter.cs

```diff
@@ -65,10 +65,8 @@ private string GetOtherPropertyDisplayName(ClientModelValidationContext context)
                 var otherPropertyDisplayName = OtherPropertyDisplayName;
                 if (otherPropertyDisplayName == null && metadata.ContainerType != null)
                 {
-                    var otherProperty = context.MetadataProvider
-                                               .GetMetadataForProperty(() => metadata.Model,
-                                                                       metadata.ContainerType,
-                                                                       OtherProperty);
+                    var containerMetadata = context.MetadataProvider.GetMetadataForType(metadata.ContainerType);
+                    var otherProperty = containerMetadata.Properties[OtherProperty];
```

Comment:
nice.  if the two properties have incompatible types, want a validation error and not an exception.  in any case `metadata.Model` in the old code wasn't used here.